### PR TITLE
[MRG+1] Fix containment test with nonlinear transforms.

### LINKED ISSUE #3540: Pick events broken in log axes
Handling pick events does not work on logarithmic axes. See the example below. As long as the line statement `ax.set_xscale('log')` is commented out, clicking on the polygons created by `axvspan` xauses them to change from gray to red. Uncomment that line to make the x axis logarithmic, and clicking on the polygons has either no effect or is misdirected to the far-left polygon only.

``` python
    from matplotlib import pyplot as plt
    fig = plt.figure()
    ax = fig.add_subplot(111)
    ax.set_xlim(1, 10)
    for i in range(1, 10, 3):
        ax.axvspan(i, i + 1, color='gray', picker=True)

    # uncomment this line to break the picker
    # ax.set_xscale('log')

    def onpick(event):
        event.artist.set_color('red')
        fig.canvas.draw()

    fig.canvas.mpl_connect('pick_event', onpick)
    plt.show()
```

Tested on version 1.4.0 and master as of 04496b056f1995e1c54e6a5feb5be9256e2d8e9b on Mac OS X Mavericks.

[TAC edited for formatting]

### LINKED ISSUE #7655: Event picking does not seem to work on polar bar plots
The minimal example provided below has been tested on Windows 10, Python 3.5.2 (Python 3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:01:18) [MSC v.1900 32 bit (Intel)] on win32 ), matplotlib 1.5.3. 

Running the code below results in events being printed for the Cartesian bar plot, but not for the polar bar plot. Changing the bar plots to scatter plots works for both Cartesian scatter plots and polar scatter plots.
```

 
#!/usr/bin/env python3
import matplotlib.pyplot as plt
 
def onpick(event):
    print(event) #Only events from fig2 are shown
    return True
rect=[.1,.1,.8,.8]
fig1 = plt.figure()
ax = fig1.add_axes(rect,polar=True)
ax.bar(1, 2,   picker=True)
fig1.canvas.mpl_connect('pick_event',onpick)

fig2 = plt.figure()
ax2 = fig2.add_axes(rect,polar=False)
ax2.bar(1, 2,  picker=True)
fig2.canvas.mpl_connect('pick_event',onpick)
plt.show()

```

### PR DESCRIPTION
**Pull request by anntzer**
Fixes #3540; see in particular https://github.com/matplotlib/matplotlib/issues/3540#issuecomment-272723725.

Edit: Also fixes #7655.

@mdboom Would it be reasonable to remove the definition of `Transform.__array__` and only keep `Affine2D.__array__`, thus forcing call sites to explicitly only pass in affine transforms to the C++ code, rather than sometimes silently dropping the nonlinear part (as in this case)?

### CODE REVIEW

**Comment by NelleV** [comment]
Can you add a regression test?

**Comment by anntzer** [comment]
Done.

File: lib/matplotlib/path.py [hunk 1]
```diff
@@ -493,9 +493,15 @@ def contains_point(self, point, transform=None, radius=0.0):
         """
         if transform is not None:
             transform = transform.frozen()
-        result = _path.point_in_path(point[0], point[1], radius, self,
-                                     transform)
-        return result
+        from .transforms import Affine2D
```
Comment (NelleV) [inline lib/matplotlib/path.py]: Any reason to have this here and not at the top of the module as python convention require?

Comment (anntzer) [inline lib/matplotlib/path.py]: Probably dates back to when I implemented this as a quickfix.  Moved up.

**Comment by anntzer** [comment]
Actually there is a circular import between transforms.py and path.py, so I left the Affine2D import inside the function, with a note to that effect.

**Comment by anntzer** [comment]
Test failure is probably related to pytest migration... we can wait until it is done, as that seems close(?).

**Comment by NelleV** [comment]
Let's wait until the pytest migration is over.

But, can we fix the circular import in a more robust way?

**Comment by anntzer** [comment]
This would likely imply a small performance cost as `transforms` relies on passing a Path to `update_path_extent` to compute not only extremal points (which is just as cheap in numpy) but also minpos (and `x[x > 0].min()` is very likely slower than a C loop).  (Fixing #7413 would probably help.)

**Comment by NelleV** [comment]
I would feel much more comfortable with fixing the circular import.

**Comment by anntzer** [comment]
There's also `transform_path_affine`, which currently reads
```
    def transform_path_affine(self, path):
        return Path(self.transform_affine(path.vertices),
                    path.codes, path._interpolation_steps)
```
and would have to become something like
```
    def transform_path_affine(self, path):
        return type(path)(self.transform_affine(path.vertices),
                    path.codes, path._interpolation_steps)
```
which is worse than a circular import IMO.

**Comment by anntzer** [comment]
What is you preferred solution other than the circular import, then?  As mentioned above I can work around it by obfuscating the implementation of `transform_path_affine` to sneakily get a handle to the Path class without a circular import, but that's not really better IMO.

File: lib/matplotlib/path.py [hunk 2]
```diff
@@ -493,9 +492,17 @@ def contains_point(self, point, transform=None, radius=0.0):
...
-                                     transform)
-        return result
+        # `point_in_path` does not handle nonlinear transforms, so we
+        # transform the path ourselves.  If `transform` is affine, letting
+        # `point_in_path` handle the transform avoids allocating an extra
+        # buffer.
+        # We need to import Affine2D here to prevent a circular import between
+        # the .path and .transforms modules.
+        from .transforms import Affine2D
+        if transform and not isinstance(transform, Affine2D):
```
Comment (tacaswell) [inline lib/matplotlib/path.py]: Can't this be `if not transform.is_affine:` ?  That seems way better that `isinstance` anyway.

Comment (anntzer) [inline lib/matplotlib/path.py]: Good catch, done.

File: lib/matplotlib/tests/test_path.py [hunk 3]
```diff
@@ -45,6 +45,23 @@ def test_contains_points_negative_radius():
...
+    assert not contains[0]
+
+
+def test_nonlinear_containment():
+    fig, ax = plt.subplots()
+    ax.set(xscale="log", ylim=(0, 1))
+    polygon = ax.axvspan(1, 10)
+    assert polygon.get_path().contains_point(
+        ax.transData.transform_point((5, .5)), ax.transData)
+
```
Comment (dstansby) [inline lib/matplotlib/tests/test_path.py]: May also be worth checking that the points (0.5, 0.5) and (20, 0.5) are not inside `polygon`

File: lib/matplotlib/path.py [hunk 4]
```diff
@@ -493,9 +492,14 @@ def contains_point(self, point, transform=None, radius=0.0):
...
         if transform is not None:
             transform = transform.frozen()
-        result = _path.point_in_path(point[0], point[1], radius, self,
-                                     transform)
-        return result
+        # `point_in_path` does not handle nonlinear transforms, so we
+        # transform the path ourselves.  If `transform` is affine, letting
+        # `point_in_path` handle the transform avoids allocating an extra
+        # buffer.
+        if transform and not transform.is_affine:
```
Comment (dstansby) [inline lib/matplotlib/path.py]: Is there a reason not to just do the transformation here even if it is an affine transformation? (thus removing the `if` block)

Comment (anntzer) [inline lib/matplotlib/path.py]: I *guess* you may save the creation of a large temporary (if the c++ code does the transform one point at a time), which would be relevant for very large paths.

To be honest I don't know but decided not to take any risks.

File: lib/matplotlib/tests/test_path.py (hunk 3 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_path.py]: done

**Pull request merged**